# Linear Regression - with Multicollinearity example

## Problem statement

Using the 3 predictor variables 'TV', 'radio', and 'newspaper' determine the 'sales'

### Multicollinearity

In regression, 'multicollinearity' refers to predictors that are correlated with other predictors. Multicollinearity occurs when a model includes multiple variables that are correlated not just with the target variable, but also with each other. In other words, it occurs with redundant variables. 

If the correlation between two predictor variables is above > 90%, it means that those two variables have a strong correlation between them. So, instead of using both the variables as predictors to predict the target variable we can can only use one of them. 

How to find the correlation betweem two variables, and
How to decide which one to choose of those two?

The following code first explores the multicollinearity and then the exercises the simple regression problem.

In [12]:
import pandas as pd

In [13]:
#import the dataset
data = pd.read_csv('Advertising.csv')

In [22]:
data

TV  radio  newspaper  sales
0    230.1   37.8       69.2   22.1
1     44.5   39.3       45.1   10.4
2     17.2   45.9       69.3    9.3
3    151.5   41.3       58.5   18.5
4    180.8   10.8       58.4   12.9
..     ...    ...        ...    ...
195   38.2    3.7       13.8    7.6
196   94.2    4.9        8.1    9.7
197  177.0    9.3        6.4   12.8
198  283.6   42.0       66.2   25.5
199  232.1    8.6        8.7   13.4

[200 rows x 4 columns]

In [23]:
#to get rid of the index column use a drop function.
data = data.iloc[:,1:]

Let's split predictor variables and the target variable. In this case "sales" is our target variable.

In [26]:
x = data[['TV','radio','newspaper']]

In [28]:
y = data[['sales']]

### One way of verifying the multicollinearity is by plotting a correlation matrix


In [36]:
import matplotlib.pyplot as plt
x.iloc[:,1:].corr()

TV     radio  newspaper
TV         1.000000  0.054809   0.056648
radio      0.054809  1.000000   0.354104
newspaper  0.056648  0.354104   1.000000

Observing the correlations amongst all the varibales, it is very less. Therefore, we can safely say there is no multicollinearity between the predictor variables.

### Another way of verifying multicollinearity is by using Ordinary Least Square model

The linear equation:

sales = Bo + B1*(TV) + B2*(radio) + B3*(newspaper)

Right now we don't have a Bo constant. Let's create a Bo variable in the dataset.

In [30]:
import statsmodels.api as sm

In [31]:
#Add the constant
x = sm.add_constant(x)

Fit an Oridinary Least Square (OLS) model to check for Multicollinearity.

In [33]:
model= sm.OLS(y, x).fit()

Now, let's check what the model has to offer.

In [34]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Wed, 23 Sep 2020   Prob (F-statistic):           1.58e-96
Time:                        18:16:42   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The constant 2.9389 is the Bo in this equation. Similarly,
B1 = 0.0458
B2 = 0.1885
B3 = -0.0010

r2 = 0.897, which is close to 1, hence a good value. (r2 is always between 0 to 1)

All Standard Error (std err) values are less. Therefore they are good as well.

All p values are less than 0.05. Only for newspaper it is 0.860 and that is beacuse the coef is negative as well. Therefore, there is no multicollinearity in the variables.

## Now let's analyse another dataset where multicollinearity exists

In [42]:
data2 = pd.read_csv('Salary_Data.csv')

Let's split predictor variables and the target variable. In this case "Salary" is our target variable.

In [45]:
X = data2.iloc[:,:2]

In [51]:
Y = data2.iloc[:,2:4]

Let's check the Multicollinearity now

In [58]:
X.iloc[:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

### Here you can see that the correlation is almost 98%

Now we know that multicollinearity exists, so we can safely drop on predictor out of those two.

But, which one?

To find out which one is safe to drop, use the OLS model to get the p-value.

In [60]:
## fit a OLS model
X = sm.add_constant(X)
model1= sm.OLS(Y, X).fit()

In [61]:
model1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Wed, 23 Sep 2020   Prob (F-statistic):           1.35e-19
Time:                        19:50:54   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We can choose the preditor with the higher p value and drop the variable.

In [14]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Sun, 29 Mar 2020   Prob (F-statistic):           1.35e-19
Time:                        23:38:56   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Now let's build a simple linear regression model

We will only use the "YearsExperience" predictor and drop the "Age" varaible.

In [69]:
#Let'split the independant and target variables. Now we are only using 'YearsExperience'

X1 = data2[['YearsExperience']]
Y1 = data2[['Salary']]

Let's split the data into train set and test set

In [72]:
from sklearn.model_selection import train_test_split

In [75]:
X1_train, X1_test, Y1_train, Y1_test = train_test_split(X1,Y1,test_size=0.2, random_state=123)

Fit the model to the train set

In [76]:
from sklearn.linear_model import LinearRegression

In [77]:
Regression = LinearRegression()

In [78]:
modelLR = Regression.fit(X1_train,Y1_train)

In [79]:
Y1_pred = modelLR.predict(X1_test)

In [80]:
from sklearn.metrics import r2_score

In [95]:
score1 = r2_score(Y1_pred, Y1_test)

In [96]:
score1

0.9770687363147101

r2 value is very close to 1. Good Simple Linear Regression model.

### Let' see what would have been the value of r2 if we had considered both 'YearsExperience' and 'Age' as predictors

In [86]:
X2 = data2[['YearsExperience','Age']]

In [87]:
Y2 = data2[['Salary']]

In [89]:
X2_train, X2_test, Y2_train, Y2_test = train_test_split(X2,Y2, test_size=0.2, random_state=123)

In [90]:
Regression2 = LinearRegression()

In [91]:
modelLR2 = Regression2.fit(X2_train,Y2_train)

In [92]:
Y2_pred = modelLR2.predict(X2_test)

In [93]:
score2 = r2_score(Y2_pred,Y2_test)

In [94]:
score2

0.9523643524150358

Here our r2 value is still close to 1, but not better than the score1. Therefore, handling multicollinearity improves the model performance.